actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -253,6 +300,57 @@ def handle_unverified_request
       end
     end
 
+    module TokenStorageMethods
+      class Base
+        attr_reader :controller
+
+        def initialize(controller)
+          @controller = controller
+        end
+
+        private
+          def generate_csrf_token # :nodoc:
+            if controller.urlsafe_csrf_tokens
+              SecureRandom.urlsafe_base64(AUTHENTICITY_TOKEN_LENGTH)
+            else
+              SecureRandom.base64(AUTHENTICITY_TOKEN_LENGTH)
+            end
+          end
+      end
+
+      class SessionStorage < Base
+        def initialize(controller)
+          super
+        end
+
+        def retrieve_token
+          controller.request.session[:_csrf_token] ||= generate_csrf_token
+        end
+      end
+
+      class CookieStorage < Base
+        attr_reader :cookie
+
+        def initialize(controller)
+          super
+        end
+
+        def retrieve_token
+          if controller.request.cookie_jar.key?(controller.csrf_token_cookie)
+            return controller.request.cookie_jar[controller.csrf_token_cookie]
+          end
+
+          csrf_token = generate_csrf_token
+          controller.request.cookie_jar[controller.csrf_token_cookie] = {
+            value: csrf_token,
+            httponly: true,
+            same_site: :lax,
```

Comment:
I'm not entirely convinced that setting `SameSite=Lax` is the correct thing to do here.

My thinking is that if `action_dispatch.cookies_same_site_protection = :none`, we don't want to use that default, because as discussed, this protection relies on the attacker not being able to get the cookie, so there really shouldn't be any case where this cookie should be sent to third-party sites.

What I'm less sure about is what we should do if `action_dispatch.cookies_same_site_protection = :strict`.  I can imagine a case where a user has a site open in one tab, and then clicks a link in another tab that redirects them to the same site.  In that case, I believe we want to share the same token between the two tabs.  If this cookie had strict protection, the cookie wouldn't be sent when clicking the link.  As such, I've opted to make this cookie always use `SameSite=Lax`.

I'm more than willing to change this if there is a flaw in my thinking.

Reply:
I think it's how it's supposed to be. We pretty much need the same setting than the session cookie.